### DISCUSSION TITLE
How to know if FFmpegFrameGrabber uses GPU for decoding mp4

### DISCUSSION THREAD

**Discussion by thhart** [bytedeco/javacv#1993] [Q&A]
I have looked around and tried to include almost everything related to cuda, I can see cuda libraries are loaded, however still I am not able to know if FFmpegFrameGrabber uses GPU or not. Is there a simple approach to check this?

I have included this at least:

       <dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>cuda-platform-redist</artifactId>
            <version>11.8-8.6-1.5.8</version>
        </dependency>

Code looks like this:

       FFmpegFrameGrabber frameGrabber = new FFmpegFrameGrabber(fileInView.getAbsolutePath());
			try {
				frameGrabber.setOption("hwaccel", "cuvid");
				frameGrabber.setOption("hwaccel_device", "0");

**Comment by saudet** [comment]
Please make sure that `FFmpegLogCallback.set()` has been called.

**Reply by thhart** [reply]
avutil.av_log_set_level(avutil.AV_LOG_TRACE);
				FFmpegLogCallback.set();

I get some kind of output there but not very useful.

Info: Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '000025.MP4':
Info:   Metadata:
Info:     major_brand     : 
Info: mp42
Info: 
Info:     minor_version   : 
Info: 0
Info: 
Info:     compatible_brands: 
Info: isomavc1mp42
Info: 
Info:     creation_time   : 
Info: 2021-02-26T13:37:43.000000Z
Info: 
Info:   Duration: 
Info: 00:10:00.00
Info: , start: 
Info: 0.000000
Info: , bitrate: 
Info: 34313 kb/s
Info: 
Info:   Stream #0:0
Info: [0x1]
Info: (eng)
Info: : Video: hevc (Main), 1 reference frame (hvc1 / 0x31637668), yuvj420p(pc, bt709, left), 3840x2160, 0/1, 33959 kb/s
Info: , 
Info: 30 fps, 
Info: 30 tbr, 
Info: 60k tbn
Info:  (default)
Info: 
Info:     Metadata:
Info:       creation_time   : 
Info: 2021-02-26T13:37:43.000000Z
Info: 
Info:       handler_name    : 
Info: 
Info: VideoHandler
Info: 
Info:       vendor_id       : 
Info: [0][0][0][0]
Info: 
Info:   Stream #0:1
Info: [0x2]
Info: (eng)
Info: : Audio: aac (LC) (mp4a / 0x6134706D), 32000 Hz, mono, fltp, 96 kb/s
Info:  (default)
Info: 
Info:     Metadata:
Info:       creation_time   : 
Info: 2021-02-26T13:37:43.000000Z
Info: 
Info:       handler_name    : 
Info: 
Info: SoundHandler
Info: 
Info:       vendor_id       : 
Info: [0][0][0][0]
Info:

**Reply by saudet** [reply]
Doesn't contain anything related to CUDA, so I guess not. Usually we need to set the codec manually with setVideoCodecName().

**Reply by thhart** [reply]
There is some interesting behaviour, when do the following:

    recorder.setVideoCodecName("h264");

I get:

    org.bytedeco.javacv.FFmpegFrameGrabber$Exception: avcodec_open2() error -22: Could not open video codec. (For more details, make sure FFmpegLogCallback.set() has been called.)

There is full log set, but there is nothing more seen than the standard output.

When I put some nonsense there:

    recorder.setVideoCodecName("foo");

I get no error whatsoever, decoding works but still I don't know if Cuda/GPU is used or not.

**Reply by saudet** [reply]
Try with the names given here:
https://trac.ffmpeg.org/wiki/HWAccelIntro#CUDANVENCNVDEC

For "hwaccel", I'd try setVideoOption(), not setOption().

**Reply by thhart** [reply]
Hi Samuel, I appreciate your input but trying things here and there is a bit frustrating. In the meantime I monitored the GPU (nvtop) while transcoding with JavaCV and it looks like it is working with the GPU. However a better way of control would be nice as well a better feedback which options are working in the underlying ffmpeg structures. I hope this can be seen as enhancement request, I really appreciate all the work in the whole framework as such and what is in there to have this level of performance.

**Reply by saudet** [reply]
Sure thing! Contributions are welcome
